src/WebJobs.Extensions.CosmosDB/Config/CosmosDBHostBuilderExtensions.cs

```diff
@@ -0,0 +1,36 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using Microsoft.Azure.WebJobs.Extensions.CosmosDB;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Microsoft.Extensions.Hosting
+{
+    /// <summary>
+    /// Extension methods for CosmosDB integration.
+    /// </summary>
+    public static class CosmosDBHostBuilderExtensions
+    {
+        internal const string ConfigSectionName = "CosmosDB";
+
+        /// <summary>
+        /// Enables use of the CosmosDB extensions
+        /// </summary>
+        /// <param name="hostBuilder">The <see cref="IHostBuilder"/> to register the extension with.</param>
+        public static IHostBuilder AddCosmosDB(this IHostBuilder hostBuilder)
+        {
+            return hostBuilder
+                .AddExtension<CosmosDBExtensionConfigProvider>()
+                .ConfigureServices(services =>
+                {
+                    services.AddSingleton<ICosmosDBServiceFactory, DefaultCosmosDBServiceFactory>();
+                    services.AddOptions<CosmosDBOptions>()
+                        .Configure<IConfiguration>((options, config) =>
+                        {
+                            config.GetSection(ConfigSectionName)?.Bind(options);
```

Comment:
If this is using `?.Bind` wouldn't it mean that in the other places where `_options` is used, it should be `_options?` since it might be null?

Reply:
Options won't be null as long as you ask for an `IOptions<CosmosDBOptions>`. By default in both the `HostBuilder` and `WebHostBuilder`, there's a call to `AddOptions()`, which registers the `OptionsManager<>` for any type of `IOptions<>`. This means anytime you request anything via `IOptions<>`, it will be instantiated (or pulled from the cache) and given to you. The call to `Bind()` will set matching properties on that object with values from `IConfiguration`, so if there's no config section found, you'll just end up with the default values.

Good question -- I had to look this up so I made sure I understood why it worked even if there was no config section :-)
- Here's where `HostBuilder` calls `AddOptions()` as part of it's call to `.Build()`: https://github.com/aspnet/Hosting/blob/master/src/Microsoft.Extensions.Hosting/HostBuilder.cs#L183
- Here's where `AddOptions()` registers the `OptionsManager<>`: https://github.com/aspnet/Options/blob/master/src/Microsoft.Extensions.Options/OptionsServiceCollectionExtensions.cs#L30
- Here's the manager: https://github.com/aspnet/Options/blob/master/src/Microsoft.Extensions.Options/OptionsManager.cs
- And the manager uses this factory for instantiating types:  https://github.com/aspnet/Options/blob/master/src/Microsoft.Extensions.Options/OptionsFactory.cs#L41